tls/extensions/s2n_client_key_share.c

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "tls/extensions/s2n_client_key_share.h"
+
+#include "crypto/s2n_ecc.h"
+#include "error/s2n_errno.h"
+#include "stuffer/s2n_stuffer.h"
+#include "utils/s2n_safety.h"
+
+
+int s2n_ecdhe_parameters_send(struct s2n_connection *conn, struct s2n_stuffer *out);
+int s2n_extensions_key_share_calculate_size();
+
+int key_share_size = -1;
+
+int s2n_extensions_client_key_share_recv(struct s2n_connection *conn, struct s2n_stuffer *extension) {
+    S2N_ERROR(S2N_ERR_UNIMPLEMENTED);
+}
+
+int s2n_extensions_client_key_share_size(struct s2n_connection *conn) {
+    if (key_share_size < 0) {
+        key_share_size = s2n_extensions_key_share_calculate_size();
+    }
+    return key_share_size;
+}
+
+int s2n_extensions_client_key_share_send(struct s2n_connection *conn, struct s2n_stuffer *out) {
+    GUARD(s2n_stuffer_write_uint16(out, TLS_EXTENSION_KEY_SHARE));
+    GUARD(s2n_stuffer_write_uint16(out, s2n_extensions_client_key_share_size(conn) - 4));
```

Comment:
is there a better way to handle these magic numbers? it'd be nice to have a function that takes a struct, write the u16 header along with its payload without to care about it's size here.